File: src/main/java/de/maggiwuerze/xdccloader/controller/DownloadController.java
package de.maggiwuerze.xdccloader.controller;

import de.maggiwuerze.xdccloader.events.SocketEvents;
import de.maggiwuerze.xdccloader.model.download.Download;
import de.maggiwuerze.xdccloader.model.download.DownloadState;
import de.maggiwuerze.xdccloader.model.entity.Bot;
import de.maggiwuerze.xdccloader.model.entity.User;
import de.maggiwuerze.xdccloader.model.entity.UserSettings;
import de.maggiwuerze.xdccloader.model.forms.DownloadForm;
import de.maggiwuerze.xdccloader.model.forms.UserForm;
import de.maggiwuerze.xdccloader.model.transport.DownloadTO;
import de.maggiwuerze.xdccloader.model.transport.UserTO;
import de.maggiwuerze.xdccloader.security.UserRole;
import de.maggiwuerze.xdccloader.service.BotService;
import de.maggiwuerze.xdccloader.service.DownloadService;
import de.maggiwuerze.xdccloader.service.EventService;
import de.maggiwuerze.xdccloader.service.UserService;
import java.security.Principal;
import java.util.Arrays;
import java.util.List;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.ui.ModelMap;
import org.springframework.validation.BindingResult;
import org.springframework.validation.Errors;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.servlet.ModelAndView;


@Controller
@Slf4j
@RequiredArgsConstructor
class DownloadController{

	private final DownloadService downloadService;
	private final BotService botService;
	private final EventService eventService;

	/**
	 * @return a list of all downloads
	 */
	@GetMapping("/downloads/")
	public ResponseEntity<List<Object>> getAllDownloads() {
		return new ResponseEntity(DownloadTO.getListOfTOs(downloadService.findAllByOrderByProgressDesc()), HttpStatus.OK);
	}

	/**
	 * @return a list of downloads. if active, then it return all that are still working. if not it returns all that have stopped, this includes errors
	 */
	@GetMapping("/downloads/active/")
	public ResponseEntity<List<Object>> getActiveDownloads(boolean active) {
		List<DownloadState> states;
		if (!active) {

			states = Arrays.asList(DownloadState.UNKNOWN, DownloadState.DONE);
		} else {

			states = Arrays.asList(DownloadState.PREPARING,
				DownloadState.PREPARED,
				DownloadState.READY,
				DownloadState.CONNECTING,
				DownloadState.TRANSMITTING,
				DownloadState.FINALIZING
			);
		}

		return new ResponseEntity(DownloadTO.getListOfTOs(downloadService.findAllByStatusInOrderByProgress(states)), HttpStatus.OK);
	}

	@GetMapping("/downloads/failed")
	public ResponseEntity<List<Object>> getFailedDownloads() {
		List<DownloadTO> failedDownloads = DownloadTO.getListOfTOs(downloadService.findAllByStatusOrderByProgressDesc(DownloadState.ERROR));

		return new ResponseEntity(failedDownloads, HttpStatus.OK);
	}

	@DeleteMapping("/downloads/")
	public ResponseEntity<?> removeDownloads(Long downloadId) {
		Download download = downloadService.getById(downloadId);

		if (download != null) {

			download.setStatus(DownloadState.STOPPED);
			download.getProgressWatcher().cancel(true);
			eventService.publishEvent(SocketEvents.DELETED_DOWNLOAD, download);

			return new ResponseEntity("Download marked for deletion", HttpStatus.OK);
		}

		return new ResponseEntity("Illegal Arguments in Request", HttpStatus.BAD_REQUEST);
	}

	@PostMapping("/downloads/")
	public ResponseEntity<?> addDownload(@RequestBody DownloadForm downloadForm) {
		Bot bot = botService.findById(downloadForm.getTargetBotId());

		if (bot == null) {
			return new ResponseEntity("Illegal Arguments in Request", HttpStatus.BAD_REQUEST);
		}

		String fileRefId = downloadForm.getFileRefId();
		if (fileRefId.contains(",")) {

			for (String id : fileRefId.split(",")) {

				Download download = new Download(bot, id);
				downloadService.addDownloadToBotQueue(download);
				eventService.publishEvent(SocketEvents.NEW_DOWNLOAD, download);
			}
		} else {
			Download download = new Download(bot, fileRefId);
			downloadService.addDownloadToBotQueue(download);
			eventService.publishEvent(SocketEvents.NEW_DOWNLOAD, download);
		}

		return new ResponseEntity("Download(s) added succcessfully.", HttpStatus.OK);
	}
}

File: src/main/kotlin/de/maggiwuerze/xdccwebloader/controller/DownloadController.kt
package de.maggiwuerze.xdccwebloader.controller

import de.maggiwuerze.xdccwebloader.events.SocketEvents.CANCELLED_DOWNLOAD
import de.maggiwuerze.xdccwebloader.events.SocketEvents.DELETED_DOWNLOAD
import de.maggiwuerze.xdccwebloader.model.download.Download
import de.maggiwuerze.xdccwebloader.model.download.DownloadState
import de.maggiwuerze.xdccwebloader.model.download.DownloadTO
import de.maggiwuerze.xdccwebloader.model.forms.DownloadFormTO
import de.maggiwuerze.xdccwebloader.service.DownloadService
import de.maggiwuerze.xdccwebloader.service.EventService
import io.swagger.v3.oas.annotations.media.Schema
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController
@RequestMapping("api/v1/download/")
class DownloadController(
    val downloadService: DownloadService,
    val eventService: EventService
) {

    @GetMapping("{id}")
    @Schema(name = "Get Download", description = "Returns a single download")
    fun getDownload(@PathVariable id: UUID): ResponseEntity<DownloadTO> = downloadService.getById(id)?.let {
        ResponseEntity(
            it.toTO(), HttpStatus.OK
        )
    } ?: ResponseEntity(HttpStatus.NOT_FOUND)


    @GetMapping
    @Schema(name = "Get Downloads", description = "Returns a list of all downloads")
    fun listDownloads(): ResponseEntity<List<DownloadTO>> = ResponseEntity(
        downloadService.findAllByOrderByProgressDesc().map { it.toTO() },
        HttpStatus.OK
    )

    /**
     * @return a list of downloads. if active, then it return all that are still working. if not it returns all that have stopped, this includes errors
     */
    @GetMapping("active/")
    @Schema(name = "Get Active Downloads", description = "Returns a list of all active downloads")
    fun getActiveDownloads(active: Boolean): ResponseEntity<List<DownloadTO>> {
        return ResponseEntity(
            when (active) {
                true -> downloadService.findAllActive()
                false -> downloadService.findAllInactive()
            }, HttpStatus.OK
        )
    }

    @GetMapping("failed")
    fun failedDownloads(): ResponseEntity<List<DownloadTO>> {
        downloadService.findAllByStatusOrderByProgressDesc(DownloadState.ERROR).map { it.toTO() }.let {
            return ResponseEntity(it, HttpStatus.OK)
        }
    }

    @DeleteMapping("{id}")
    fun removeDownload(@PathVariable id: UUID): ResponseEntity<*> {
        return downloadService.getById(id)?.let { download ->
            download.progressWatcher?.cancel(true)
            download.status = DownloadState.STOPPED
            eventService.publishEvent(DELETED_DOWNLOAD, download)

            ResponseEntity("Download marked for deletion", HttpStatus.OK)
        } ?: ResponseEntity("Download not found", HttpStatus.NOT_FOUND)
    }

    @PostMapping
    fun addDownload(@RequestBody downloadFormTO: DownloadFormTO): ResponseEntity<List<DownloadTO>> {
        return ResponseEntity(downloadService.create(downloadFormTO).map(Download::toTO), HttpStatus.OK)
    }

    @PutMapping("{id}")
    fun cancelDownload(@PathVariable id: UUID): ResponseEntity<*> {
        return downloadService.getById(id)?.let { download ->
            download.progressWatcher?.cancel(true)
            download.status = DownloadState.STOPPED
            eventService.publishEvent(CANCELLED_DOWNLOAD, download)

            ResponseEntity("Download marked for deletion", HttpStatus.OK)
        } ?: ResponseEntity("Download not found", HttpStatus.NOT_FOUND)
    }
}